test/unit/event/src/EventTest.cpp

```diff
@@ -227,44 +236,260 @@ TEMPLATE_LIST_TEST_CASE("waitForEventThatAlreadyFinishedShouldBeSkipped", "[even
+            REQUIRE(!alpaka::isComplete(k1_0));
+            REQUIRE(!alpaka::isComplete(k1_1));
+            REQUIRE(alpaka::isComplete(k2));
+            REQUIRE(alpaka::isComplete(k3));
+            REQUIRE(!alpaka::isComplete(e1));
+            REQUIRE(!alpaka::isComplete(e2));
+            REQUIRE(!alpaka::isComplete(e3));
+
+            // After the kernel k1_0 is released e3 is not allowed to be ready because q3 depends on the oldest e1_new
+            // state.
+            k1_0.trigger();
+            // q1 = [k1_1,e1_new]
+            // q2 = []
+            // q3 = [->e1_new,e3]
+
+            REQUIRE(alpaka::isComplete(k1_0));
+            REQUIRE(!alpaka::isComplete(k1_1));
+            REQUIRE(alpaka::isComplete(k2));
+            REQUIRE(alpaka::isComplete(k3));
+            REQUIRE(!alpaka::isComplete(e1));
+            REQUIRE(alpaka::isComplete(e2));
+            REQUIRE(!alpaka::isComplete(e3));
+
+            k1_1.trigger();
+            // q1 = []
             // q2 = []
+            REQUIRE(alpaka::isComplete(k1_0));
+            REQUIRE(alpaka::isComplete(k1_1));
+            REQUIRE(alpaka::isComplete(k2));
+            REQUIRE(alpaka::isComplete(k3));
+            REQUIRE(alpaka::isComplete(e1));
+            REQUIRE(alpaka::isComplete(e2));
+            REQUIRE(alpaka::isComplete(e3));
+        }
+        else
+        {
+            std::cerr << "Can not execute test because CU_DEVICE_ATTRIBUTE_CAN_USE_STREAM_MEM_OPS is not supported!"
+                      << std::endl;
+        }
+    }
+}
+
+TEMPLATE_LIST_TEST_CASE("eventReEnqueueWithSomeoneWaitsForEventOutOfOrderLifetimeRelease", "[event]", TestQueues)
+{
+    // A re-enqueued event will be released in the opposite order it is recorded.
+    // The tests validate that dependencies between queues and the re-enqueued event will be correct.
+    using DevQueue = TestType;
+    using Fixture = alpaka::test::QueueTestFixture<DevQueue>;
+    using Queue = typename Fixture::Queue;
+    using Dev = typename Fixture::Dev;
+
+    if(!alpaka::test::IsBlockingQueue<Queue>::value)
+    {
+        Fixture f1;
+        Fixture f2;
+        Fixture f3;
+        if(alpaka::test::isEventHostManualTriggerSupported(f1.m_dev)
+           && alpaka::test::isEventHostManualTriggerSupported(f2.m_dev))
+        {
+            auto q1 = f1.m_queue;
+            auto q2 = f2.m_queue;
+            auto q3 = f3.m_queue;
+            alpaka::Event<Queue> e1(f1.m_dev);
+            alpaka::Event<Queue> e2(f2.m_dev);
+            alpaka::Event<Queue> e3(f3.m_dev);
+            alpaka::test::EventHostManualTrigger<Dev> k1_0(f1.m_dev);
+            alpaka::test::EventHostManualTrigger<Dev> k2(f2.m_dev);
+
+            alpaka::enqueue(q1, k1_0);
+            alpaka::enqueue(q1, e1);
+            // q1 = [k1_0,e1]
+            alpaka::enqueue(q2, k2);
+            // q2 = [k2]
+            REQUIRE(!alpaka::isComplete(k1_0));
+
+            alpaka::wait(q3, e1); // let wait
```

Comment:
```suggestion
            alpaka::wait(q3, e1);
```

Looks like a leftover from our VC ;-)

Reply:
Yes, note in PIConGPU we call it `waitOn`